metaswitch/common/alarms.py

```diff
@@ -31,36 +31,281 @@
+        """Send or re-send the alarm's state to the alarm agent."""
+        self._last_state_raised.issue()
+
+
+class Alarm(BaseAlarm):
+    """Alarm with only a single non-cleared severity.
+
+    The parameter severity should be passed a severity constant from this
+    module"""
+    def __init__(self, issuer, index, severity):
+        super(Alarm, self).__init__(issuer, index)
+        self._alarm_state = AlarmState(issuer, index, severity)
+
+    def set(self):
+        """Send the alarm's raised state to the alarm agent."""
+        self._last_state_raised = self._alarm_state
+        self.re_sync()
+
+
+class MultiSeverityAlarm(BaseAlarm):
+    """Alarm with multiple possible non-cleared severities.
+
+    The parameter severities should be passed an iterable of severity
+    constants from this module.
+    """
+    def __init__(self, issuer, index, severities):
+        super(MultiSeverityAlarm, self).__init__(issuer, index)
+        assert len(severities) >= 2, 'Must have multiple severities.'
+        assert CLEARED not in severities, 'Severities must not include CLEARED.'
+        self._severities = {severity: AlarmState(issuer, index, severity) for
+                            severity in severities}
+
+    def set(self, severity):
+        """Send one of the alarm's raised states to the alarm agent.
+
+        The severity must be one of the severity constants defined in this
+        module. If this alarm cannot be raised with that severity, a KeyError
+        is raised."""
+        try:
+            self._last_state_raised = self._severities[severity]
+        except KeyError:
+            _log.error('Attempted to raise incorrect alarm state %s',
+                       severity)
+            raise
+
+        self.re_sync()
+
+
+class AlarmState(object):
+    """One of an alarm's possible states."""
+    def __init__(self, issuer, index, severity):
+        self.issuer = issuer
+        self.index = index
+        self.severity = severity
+
+    def issue(self):
+        """Tell the alarm agent that this is the current state of the alarm."""
+        identifier = '{}.{}'.format(self.index, self.severity)
+        _issue_alarm(self.issuer, identifier)
+
+
+def _issue_alarm(process, identifier):
+    """Attempt to send an alarm to the alarm agent.
+
+    This function will time out after 2s.
+    """
     # Clearwater nodes all have clearwater-infrastructure installed.
     # It includes a command-line script that can be used to issue an alarm.
     # We import the function used by the script and re-use it.
     #
     # See https://github.com/Metaswitch/clearwater-infrastructure/blob/master/clearwater-infrastructure/usr/share/clearwater/bin/alarms.py
     # for the target module.
-    global sendrequest
-    if sendrequest is None:
-        try:
+    global _sendrequest
+    if _sendrequest is None:
+        try: # pragma: no cover
+            # We can't rely on this file being present in unit tests so don't
```

Comment:
This code was previously untested, and I decided that it would be more trouble than it's worth to change that. Mocking out _sendrequest isn't particularly helpful - the interesting part of this code is how it interacts with the import internals to get the function it wants, and how it caches the result. I'd test this by mocking out imp. You'd also need to be careful as this function manipulates global state; we'd have to be careful about running tests in parallel.

So the answer to your implied question is that yes, we can test this function, but I don't think it's required for this story. If you agree, I'll raise an issue to cover the missing coverage.